creusot-contracts/src/resource.rs

```diff
@@ -0,0 +1,226 @@
+pub mod fmap_view;
+
+#[cfg(creusot)]
+use crate::logic::ra;
+use crate::{
+    Ghost, Snapshot,
+    logic::{Id, Set, ra::RA},
+    *,
+};
+use ::std::marker::PhantomData;
+
+/// A ghost wrapper around a [resource algebra](RA).
+///
+/// This structure is meant to be manipulated in [`ghost`](mod@ghost) code. It is
+/// guaranteed to always contain a [`valid`](RA::valid) resource.
+///
+/// The usual usage is this:
+/// - [Create](Self::alloc) some ghost resource
+/// - [Split](Self::split) it into multiple parts, some of which may be [duplicable](RA::idemp)
+/// - [Join](Self::join) these parts later. By the validity invariant, this allows
+///   one to learn information about one part from the other.
+///
+/// # Example
+///
+/// ```rust
+/// use creusot_contracts::{*, resource::Resource, logic::ra::Ag};
+/// let mut res: Ghost<Resource<Ag<Int>>> = Resource::alloc(snapshot!(Ag::Ag(1)));
+///
+/// ghost! {
+///     let part = res.split_off(snapshot!(Ag::Ag(1)), snapshot!(Ag::Ag(1)));
+///     // Pass `part` around, forget what it contained...
+///     let _ = res.join_shared(&part);
+///     // And now we remember: the only way this works is if `part` contained `1`!
+///     proof_assert!(part@ == Ag::Ag(1));
+/// };
+/// ```
+pub struct Resource<R>(PhantomData<R>);
+
+impl<R: RA> View for Resource<R> {
+    type ViewTy = R;
+    #[logic]
+    #[open]
+    fn view(self) -> R {
+        self.val()
+    }
+}
+
+impl<R: RA> Invariant for Resource<R> {
+    #[predicate]
+    #[open]
+    fn invariant(self) -> bool {
+        self.val().valid()
+    }
+}
+
+#[allow(unused_variables)]
+impl<R: RA> Resource<R> {
+    /// Get the id for this resource.
+    ///
+    /// This prevents mixing resources of different origins.
+    #[logic]
+    #[trusted]
+    pub fn id(self) -> Id {
+        dead
+    }
+
+    /// Get the id for this resource.
+    ///
+    /// This is the same as [`Self::id`], but for ghost code.
+    #[pure]
+    #[trusted]
+    #[ensures(result == self.id())]
+    pub fn id_ghost(&self) -> Id {
+        panic!("ghost code only")
+    }
+
+    /// Get the RA contained in this resource.
+    #[logic]
+    #[trusted]
+    pub fn val(self) -> R {
+        dead
+    }
+
+    /// Create a new resource from a valid value.
+    ///
+    /// # Corresponding reasoning
+    ///
+    /// `⌜valid(value)⌝ ⊢ ∃γ, Own(value, γ)`
+    #[trusted]
+    #[requires(r.valid())]
+    #[ensures(result@ == *r)]
+    pub fn alloc(r: Snapshot<R>) -> Ghost<Self> {
+        Ghost::conjure()
+    }
```

Comment:
```suggestion
    #[trusted]
    #[pure]
    #[requires(r.valid())]
    #[ensures(result@ == *r)]
    pub fn alloc(r: Snapshot<R>) -> Ghost<Self> {
        Ghost::conjure()
    }
```

Uh oh, the purity checker should have spotted this. Is there a bug in the purity checker?

Reply:
So we never ever call this function in ghost context? Including in your persistant array code?